File: client/Classes/Basic/LuckySprite.cpp
#include "LuckySprite.h"
#include "DictDataManager.h"

USING_NS_CC;

LuckySprite::LuckySprite(void)
{
    m_nImageId =  0;
}

LuckySprite::LuckySprite(int imageId)
{
    m_nImageId = imageId;
}

LuckySprite::~LuckySprite(void)
{
}

bool LuckySprite::init()
{
    const stImage *pImage = DictDataManager::shareDictDataManager()->getImage(m_nImageId);
    if (!pImage) 
    {
        return false;
    }
    
    if (pImage->type == 0) 
    {
        initWithFile(pImage->filePath.c_str());
    }
    else 
    {
        initWithSpriteFrameName(pImage->name.c_str());
    }
    
    return true;
}

LuckySprite* LuckySprite::create(int imageId, float scale)
{
    LuckySprite *pSprite = new LuckySprite(imageId);
    if (pSprite && pSprite->init())
    {
        pSprite->autorelease();
        return pSprite;
    }
    CC_SAFE_DELETE(pSprite);
    return NULL;
}
//
//cocos2d::CCTexture2D *LuckySprite::getTextrue2D(const char *file_name)
//{
//    return cocos2d::CCTextureCache::sharedTextureCache()->addImage(file_name);
//}

//cocos2d::CCSpriteFrame *LuckySprite::getSpriteFrame(const char *_name)
//{
//    return cocos2d::CCSpriteFrameCache::sharedSpriteFrameCache()->spriteFrameByName(_name);
//}

bool LuckySprite::setNewTexture(int imageId)
{
    if (imageId == m_nImageId) 
    {
        return true;
    }
    const stImage *pImage = DictDataManager::shareDictDataManager()->getImage(imageId);
    if (!pImage) 
    {
        return false;
    }
    
    m_nImageId = imageId;

    if (pImage->type == 0) 
    {
        cocos2d:CCTexture2D *pTexture = CCTextureCache::sharedTextureCache()->addImage(pImage->filePath.c_str());
        if (!pTexture) 
        {
            return false;
        }
        setTexture(pTexture);
    }
    else 
    {
        CCSpriteFrame *pFrame = CCSpriteFrameCache::sharedSpriteFrameCache()->spriteFrameByName(pImage->name.c_str());
        if (!pFrame) 
        {
            return false;
        }
        setDisplayFrame(pFrame);
    }    
    return true;
}

File: client/Classes/Basic/LuckySprite.js
var lc = lc = lc || {};

lc.LuckySprite = cc.Sprite.extend({
    m_nImageId : 0,
    m_fScale : 1.0,
    ctor:function () {
        this._super();
    },
    //在这里初始化数据,读取数据库
    init:function () {
        this._super();

        if (this.m_nImageId > 0)
        {
            var pImage = lc.DictDataManager.getShareInstance().getImage(this.m_nImageId);
            if ( !pImage )
            {
                return false;
            }

            if (pImage.type == 0)
            {
                this.initWithFile(pImage.filePath);
            }
            else
            {
                this.initWithSpriteFrameName(pImage.name);
            }

            return true;
        }

        return false;
    },
    setNewTexture:function (imageId)
    {
        if (imageId == m_nImageId)
        {
            return true;
        }

        var pImage = lc.DictDataManager.getShareInstance().getImage(this.m_nImageId);
        if (!pImage)
        {
            return false;
        }

        m_nImageId = imageId;

        if (pImage.type == 0)
        {
            var pTexture = cc.SpriteFrameCache.getInstance().getSpriteFrame(pImage.filePath);
            if (!pTexture)
            {
                return false;
            }
            this.setTexture(pTexture);
        }
        else
        {
            var pFrame = cc.SpriteFrameCache.getInstance().getSpriteFrame(pImage.name);
            if (!pFrame)
            {
                return false;
            }
            this.setDisplayFrame(pFrame);
        }

        return true;
    }
});

lc.LuckySprite.create = function (imageId, scale)
{
    var retValue = new lc.LuckySprite();
    switch (arguments.length) {
        case 2:
        {
            retValue.m_nImageId = imageId;
            retValue.m_fScale = scale;
            break;
        }
        case 1:
        {
            retValue.m_nImageId = imageId;
            break;
        }
        default:
            break;
    }

    if ( retValue && retValue.init() )
    {
        return retValue;
    }
    return null;
};